### ISSUE TITLE
No database found when the key database is locked.

### ISSUE THREAD

**Issue by raphaelahrens** [Frederick888/git-credential-keepassxc#14]
I configured keepassxc to automagically lock the database after a period of inactivity.

So usually when I do a `git push` the database is locked and I get the follwoing error from git-credentials-keepassxc.

> ERRO No valid database associations found in configuration file, Caused by: N/A, Message: No valid database associations found in configuration file

Since the browser plugin is able to trigger keepassxc to request an unlocking of the DB, I figured that this must be possible to with git-credentials-keepassxc too. I looked in the source of the browser plugin and found that they set somewhere the flag `triggerUnlock` which then gets build into the request in keepass.js:1215. https://github.com/keepassxreboot/keepassxc-browser/blob/develop/keepassxc-browser/background/keepass.js#L1215

The argument `triggerUnlock` can be set by a call to `keepass.testAssociation`, so it could be that they do it with the `testAssociation` message/request. But I'm not sure, the JS code is very difficult to read with all the default arguments, which are forwarded seven layers deep.

Long story short it would be a nice feature, to be able to run git-credentials-keepassxc with a locked DB and then be asked to unlock it.

**Comment by Frederick888** [comment]
Thanks for digging into their code! I tested the parameter and it did work, and it's now in `issue-14` branch if you'd like to try it out. 

However one problem I noticed was that KPXC actually does _not_ block the client until the database is unlocked in this case. I still get `Database not opened` immediately. The browser extension handles this by [checking the database status once per second](https://github.com/keepassxreboot/keepassxc-browser/blob/6c8cb7db607a3099592c0d8feb7e0b96bd6b0c9a/keepassxc-browser/background/keepass.js#L1124), but `git-credential-keepassxc` hasn't got a daemon sitting in the background to do this.

It's probably better to have some new CLI arguments e.g. `--retry <-1|N> --interval <MILLISECONDS>`. Hopefully I can figure out some time to do this in the next few days.

**Comment by szaffarano** [comment]
Hi @Frederick888 I've just tried this feature and it works fine! I'm looking forward to the `--retry` argument to improve the little problem with the non-blocking issue.
Thanks a lot!

**Comment by raphaelahrens** [comment]
Hi,

I tried this as well and it indeed triggers the password prompt.

But it's strange that KPXC is not able to block the request.
The `--retry` flag  will bring some usability problems with it, for example if a user takes to long to enter the password git-credential-keepassxc will fail.

Are you able to check if the DB is locked?

In the Logs I found 

> DEBG` RECV: {"action":"test-associate","error":"Database not opened","errorCode":"1"}

which could mean that the DB is locked, so this could be used to test if the DB is still looked and one could retry until is was opened or the user kills `git-credential-keepassxc`.

But this is speculation on my part. Is there a proper documentation of the protocol?

**Comment by Frederick888** [comment]
> The `--retry <N>` flag will bring some usability problems with it, for example if a user takes to long to enter the password git-credential-keepassxc will fail.

I actually don't think that'll be an issue. It's easy enough to have a special number, e.g. 0 or -1, which allows indefinite wait.

> Are you able to check if the DB is locked?

IIRC the `get-databasehash` can do this. haven't got time to check it out (been a tad lazy recently lol).

**Comment by Frederick888** [comment]
Still on `issue-14` branch. Now there is a `--unlock` option:
```sh
$ git-credential-keepassxc --help
...
OPTIONS:
...
        --unlock <unlock>    Try unlocking database, applies to get, store and erase only.
                             Takes one argument in the format of [<MAX_RETRIES>[,<INTERVAL_MS>]]. Use 0 to retry
                             indefinitely. The default interval is 1000ms.
...
```

Usage for example, `printf 'url=https://example.com\n' | git-credential-keepassxc --unlock "" get` or `printf 'url=https://example.com\n' | git-credential-keepassxc --unlock "20,2000" get`.

I'd later manually test it for a few more times before merging the branch.

**Comment by szaffarano** [comment]
Hi @Frederick888 it worked very well, both with max retries or indefinitely retries and as git credential helper and as a standalone script as well.  Thanks!!

**Comment by stdedos** [comment]
Notable "side effect": Canceling the authentication dialog will not abort `git-credential-keepassxc`

Ofc, there might not be a way for you to detect that ... 😞
